src/kv/test/kv_dynamic_tables.cpp

```diff
@@ -0,0 +1,292 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the Apache 2.0 License.
+#include "kv/store.h"
+#include "kv/test/null_encryptor.h"
+
+#include <doctest/doctest.h>
+
+struct MapTypes
+{
+  using StringString = kv::Map<std::string, std::string>;
+  using NumNum = kv::Map<size_t, size_t>;
+  using NumString = kv::Map<size_t, std::string>;
+  using StringNum = kv::Map<std::string, size_t>;
+};
+
+TEST_CASE("Basic dynamic table" * doctest::test_suite("dynamic"))
+{
+  kv::Store kv_store;
+
+  auto encryptor = std::make_shared<kv::NullTxEncryptor>();
+  kv_store.set_encryptor(encryptor);
+
+  constexpr auto map_name = "mapA";
+
+  INFO("Dynamically created maps can be used like normal maps");
+
+  {
+    auto map_a = kv_store.get<MapTypes::StringString>(map_name);
+    REQUIRE(map_a == nullptr);
+  }
+
+  {
+    auto tx = kv_store.create_tx();
+
+    auto view = tx.get_view2<MapTypes::StringString>(map_name);
+    view->put("foo", "bar");
+
+    REQUIRE(tx.commit() == kv::CommitSuccess::OK);
+  }
+
+  {
+    INFO("Old style access");
+    auto map_a = kv_store.get<MapTypes::StringString>(map_name);
+    REQUIRE(map_a != nullptr);
+
+    auto tx = kv_store.create_tx();
+
+    auto view = tx.get_view(*map_a);
+    const auto it = view->get("foo");
+    REQUIRE(it.has_value());
+    REQUIRE(it.value() == "bar");
+  }
+
+  {
+    INFO("New style access");
+    auto tx = kv_store.create_tx();
+
+    auto view = tx.get_view2<MapTypes::StringString>(map_name);
+    const auto it = view->get("foo");
+    REQUIRE(it.has_value());
+    REQUIRE(it.value() == "bar");
+  }
+
+  {
+    INFO("Dynamic tables remain through compaction");
+    kv_store.compact(kv_store.current_version());
+
+    auto map_a = kv_store.get<MapTypes::StringString>(map_name);
+    REQUIRE(map_a != nullptr);
+
+    auto tx = kv_store.create_tx();
+
+    auto view = tx.get_view(*map_a);
+    const auto it = view->get("foo");
+    REQUIRE(it.has_value());
+    REQUIRE(it.value() == "bar");
+  }
+
+  const auto version_before = kv_store.current_version();
+
+  {
+    INFO("Multiple dynamic tables can be created in a single tx");
+    auto tx = kv_store.create_tx();
+
+    auto [v1, v2] =
+      tx.get_view2<MapTypes::StringString, MapTypes::StringNum>("1", "2");
```

Comment:
Nitpick: Could the tables be named "mapX" or something similar as "1" and "2" look more like keys and values.